pkg/admission/cert/writer/fs.go

```diff
@@ -16,125 +16,175 @@ limitations under the License.
+	certgenerator "sigs.k8s.io/controller-runtime/pkg/admission/cert/generator"
+	"sigs.k8s.io/controller-runtime/pkg/admission/cert/writer/internal/atomic"
+)
+
+const (
+	// FSCertProvisionAnnotationKeyPrefix should be used in an annotation in the following format:
+	// fs.certprovisioner.kubernetes.io/<webhook-name>: path/to/certs/
+	// the webhook cert manager library will provision the certificate for the webhook by
+	// storing it under the specified path.
+	// format: fs.certprovisioner.kubernetes.io/webhookName: path/to/certs/
+	FSCertProvisionAnnotationKeyPrefix = "fs.certprovisioner.kubernetes.io/"
+)
+
+// FSCertWriter provisions the certificate by reading and writing to the filesystem.
+type FSCertWriter struct {
+	CertGenerator certgenerator.CertGenerator
+}
+
+var _ CertWriter = &FSCertWriter{}
+
+// EnsureCerts provisions certificates for a webhook configuration by writing them in the filesystem.
+func (s *FSCertWriter) EnsureCerts(webhookConfig runtime.Object) error {
+	webhookMap := map[string]*webhookAndPath{}
+
+	accessor, err := meta.Accessor(webhookConfig)
+	if err != nil {
+		return err
+	}
+	annotations := accessor.GetAnnotations()
+	if annotations == nil {
+		return nil
+	}
+
+	// Parse the annotations to extract info
+	for k, v := range annotations {
+		// an example annotation: fs.certprovisioner.kubernetes.io/webhookName: path/to/certs/
+		if strings.HasPrefix(k, FSCertProvisionAnnotationKeyPrefix) {
+			webhookName := strings.TrimPrefix(k, FSCertProvisionAnnotationKeyPrefix)
+			webhookMap[webhookName] = &webhookAndPath{
+				path: v,
+			}
+		}
+	}
+
+	webhooks, err := getWebhooksFromObject(webhookConfig)
+	if err != nil {
+		return err
+	}
+	for i, webhook := range webhooks {
+		if p, found := webhookMap[webhook.Name]; found {
+			p.webhook = &webhooks[i]
+		}
+	}
+
+	// validation
+	for k, v := range webhookMap {
+		if v.webhook == nil {
+			return fmt.Errorf("expecting a webhook named %q", k)
+		}
+	}
+
+	generator := s.CertGenerator
+	if s.CertGenerator == nil {
+		generator = &certgenerator.SelfSignedCertGenerator{}
+	}
+
+	cw := &fsCertWriter{
+		certGenerator: generator,
+		webhookConfig: webhookConfig,
+		webhookMap:    webhookMap,
+	}
+	return cw.ensureCert()
+}
+
+// fsCertWriter deals with writing to the local filesystem.
+type fsCertWriter struct {
+	certGenerator certgenerator.CertGenerator
+
+	webhookConfig runtime.Object
+	webhookMap    map[string]*webhookAndPath
+}
+
+type webhookAndPath struct {
+	webhook *admissionregistrationv1beta1.Webhook
+	path    string
+}
+
+var _ certReadWriter = &fsCertWriter{}
+
+func (s *fsCertWriter) ensureCert() error {
+	var err error
+	for _, v := range s.webhookMap {
+		err = handleCommon(v.webhook, s)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (f *fsCertWriter) write(webhookName string) (*certgenerator.Artifacts, error) {
+	return f.doWrite(webhookName)
+}
+
+func (f *fsCertWriter) overwrite(webhookName string) (*certgenerator.Artifacts, error) {
```

Comment:
`overwrite` is the same as `write`. Can we remove it?

Reply:
It's part of the interface.
```
var _ certReadWriter = &fsCertWriter{}
```
https://github.com/kubernetes-sigs/controller-runtime/pull/57/files#diff-f49e01505d7594a4be384c43db7a00e8R120